File: Data Structures & Algorithms/max-area-of-island/submission-0.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int maxArea = 0;

        for(int r = 0; r < grid.length; r++) {
            for(int c = 0; c < grid[0].length; c++) {
                if(grid[r][c] == 1) {
                    maxArea = Math.max(maxArea, dfs(r, c, grid));
                }
            }    
        }
        return maxArea;
    }

    public int dfs(int r, int c, int[][] grid) {
        if(r < 0 || c < 0 || r >= grid.length || c >= grid[0].length || grid[r][c] != 1) return 0;

        grid[r][c] = -1;

        return 1 + dfs(r, c + 1, grid) + dfs(r, c - 1, grid) + dfs(r - 1, c, grid) + dfs(r + 1, c, grid);
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = [] # (i, height)
        maxArea = 0

        for i, height in enumerate(heights):
            start = i

            while stack and stack[-1][1] > height:
                index, h = stack.pop()
                maxArea = max(maxArea, h * (i - index))
                start = index
            stack.append([start, height])

        for i, height in stack:
            maxArea = max(maxArea, height * (len(heights) - i))

        return maxArea